### PR TITLE
Return None value during style validation, before running into exception

### PR DESCRIPTION
**Pull request by stantonxu**
This is fixing https://github.com/pybee/travertino/issues/3 

If the style value equals to 'none', set it to None and return it, before the `validate` function running into exception.

@freakboy3742 , could you please review?

### CODE REVIEW

**Comment by stantonxu** [comment]
@freakboy3742 , it looks like the smoke test is expecting an exception when the value is set to 'none'. I don't quite understand yet. Why?

File: travertino/declaration.py [hunk 1]
```diff
@@ -50,6 +50,7 @@ def validate(self, value):
                 pass
         if value == 'none':
             value = None
+            return value
```
Comment (rayrrr) [inline travertino/declaration.py]: Instead of doing `return value` here, try removing:

```
        if value == 'none':
             value = None
```

See if that gets the tests to pass.

Comment (stantonxu) [reply travertino/declaration.py]: @rayrrr , thanks. I've already tried that, doesn't work.

When I comment out the two lines,
```
        if value == 'none':
             value = None
```

The test cases of 'test_multiple_choices', 'test_none', 'test_values' failed.

Comment (rayrrr) [reply travertino/declaration.py]: Too bad. I'll give it some more thought too and see if I can help you get this done.

Comment (rayrrr) [reply travertino/declaration.py]: @stantonxu you should be able to fix this without changing any existing tests by removing the `return value` line, and changing the line `if value == 'none':` to (pseudocode) `if value == 'none' and <type is Color, Integer, or Number>:`. Try that.

Comment (stantonxu) [reply travertino/declaration.py]: The `validate` code is as below, so the `<type is Color, Integer, or Number>:` code won't hit.

```
    def validate(self, value):
        if self.default:
            if value is None:
                return None
        if self.string:
            try:
                return value.strip()
            except AttributeError:
                pass
        if self.integer:
            try:
                return int(value)
            except (ValueError, TypeError):
                pass
        if self.number:
            try:
                return float(value)
            except (ValueError, TypeError):
                pass
        if self.color:
            try:
                return color(value)
            except ValueError:
                pass
        if value == 'none':
            value = None
            return value
        for const in self.constants:
            if value == const:
                return const

        raise ValueError("'{0}' is not a valid initial value".format(value))

```

Comment (rayrrr) [reply travertino/declaration.py]: What we want to do is change `'none'` to `None` for Color, Integer, and Number types. Let's do that _before_ the conditional checks for those types.

```
 def validate(self, value):
        if self.default:
            if value is None:
                return None
        if value == 'none' and (self.color or self.integer or self.number):
            value = None
        if self.string:
            try:
                return value.strip()
            except AttributeError:
                pass
        if self.integer:
            try:
                return int(value)
            except (ValueError, TypeError):
                pass
        if self.number:
            try:
                return float(value)
            except (ValueError, TypeError):
                pass
        if self.color:
            try:
                return color(value)
            except ValueError:
                pass
        for const in self.constants:
            if value == const:
                return const

        raise ValueError("'{0}' is not a valid initial value".format(value))
```

Also, notice the lines near the end of this block starting with `for const in self.constants`. It means that for this to avoid raising a `ValueError`, the value needs to appear in the `constants.py` file. We need to move some constants from Toga to Travertino and then import them from the latter to the former.

Comment (rayrrr) [reply travertino/declaration.py]: @stantonxu once pybee/travertino#5 and pybee/toga#564 are merged, the code block above should work. Do you understand?

Comment (rayrrr) [reply travertino/declaration.py]: Actually, the other PR's might not be needed, but I still think they'll make these apps more tidy. Try the code block as is and see what happens.

Comment (stantonxu) [reply travertino/declaration.py]: @rayrrr thanks for the suggestion, however it doesn't work out. Here is the error from test,
```
======================================================================
ERROR: test_none (tests.test_choices.PropertyChoiceTests)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/Users/xus12/git/pybee/travertino/travertino/declaration.py", line 181, in setter
    value = choices.validate(value)
  File "/Users/xus12/git/pybee/travertino/travertino/declaration.py", line 60, in validate
    raise ValueError("'{0}' is not a valid initial value".format(value))
ValueError: 'none' is not a valid initial value

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/Users/xus12/git/pybee/travertino/tests/test_choices.py", line 46, in test_none
    obj.prop = 'none'
  File "/Users/xus12/git/pybee/travertino/travertino/declaration.py", line 184, in setter
    value, name, choices
ValueError: Invalid value 'none' for property 'prop'; Valid values are: none

======================================================================
ERROR: test_values (tests.test_choices.PropertyChoiceTests)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/Users/xus12/git/pybee/travertino/travertino/declaration.py", line 181, in setter
    value = choices.validate(value)
  File "/Users/xus12/git/pybee/travertino/travertino/declaration.py", line 60, in validate
    raise ValueError("'{0}' is not a valid initial value".format(value))
ValueError: 'none' is not a valid initial value

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/Users/xus12/git/pybee/travertino/tests/test_choices.py", line 260, in test_values
    obj.prop = 'none'
  File "/Users/xus12/git/pybee/travertino/travertino/declaration.py", line 184, in setter
    value, name, choices
ValueError: Invalid value 'none' for property 'prop'; Valid values are: a, b, none

----------------------------------------------------------------------
Ran 84 tests in 0.038s

FAILED (errors=2)
```

**Comment by rayrrr** [comment]
@freakboy3742 I'm pretty sure we need to change the unit tests here to reflect this need (based on the choices for a `Box` object: (`visible`, `hidden`, and `none`). 

Right now the tests assume every value of `none` will be changed to a Python `None` object, but we only want to do that for certain types (color, integer, and number). Can we change the tests accordingly, to remove the tests that make sure `none` always gets changed to `None`?

**Comment by stantonxu** [comment]
@freakboy3742 , do you have any comments? I think we probably don't need to update the test cases, but am struggling where the problem is with the code which prevent the testing from successful.

**Comment by freakboy3742** [comment]
@stantonxu @rayrrr I don't think the tests need to be altered (or I'm misunderstanding which test you're suggesting needs to be changed). The test_allow_color test, for example, validates that a property whose choices allow for Color options don't allow anything else. The test validates (and, int this patch, fails) when provided `'none'` as a value - which is the correct behavior.

**Comment by stantonxu** [comment]
@freakboy3742 , I agree with you, I think the test cases are valid. However, for the ‘Validate’ method, anysuggestion how to make it work?

**Comment by rayrrr** [comment]
@freakboy3742 and @stantonxu it was a bit tough to explain what tests I thought should be changed so I made #6 which has code changes along with passing tests and should solve the original `Box` issue. So if you agree, we can just merge it after code review. Otherwise we can use it as a reference point for further discussion here.

**Comment by phildini** [comment]
Hi there! It looks like this PR might be dead, so we're closing it for now. Feel free to re-open it if you'd like to continue, or think about directing your efforts to https://github.com/beeware/briefcase or https://github.com/beeware/toga. Both of these have more active development right now. 😄
